### PR TITLE
PLFM-1890

### PR DESCRIPTION
**Pull request by brucehoff**

### CODE REVIEW

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/RestrictableObjectDescriptor.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+ 	"properties":{
+        "id":{
+            "type":"string",
+            "transient":true
```
Comment (john-hill) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/RestrictableObjectDescriptor.json]: remove transient

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/migration/MigrationType.json [hunk 2]
```diff
@@ -16,7 +16,8 @@
         "ACL_ACCESS_TYPE",
         "FAVORITE",
         "ACCESS_REQUIREMENT",
-        "NODE_ACCESS_REQUIRMENT",
+        "NODE_ACCESS_REQUIREMENT",
+        "EVALUATION_ACCESS_REQUIREMENT",
```
Comment (john-hill) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/migration/MigrationType.json]: need to be after evaluations. The order should match the migration order.

File: services/repository-managers/src/main/java/org/sagebionetworks/evaluation/manager/EvaluationManagerImpl.java [hunk 3]
```diff
@@ -122,22 +123,12 @@ public boolean isEvalAdmin(UserInfo userInfo, String evalId) throws DatastoreExc
 		EvaluationUtils.ensureNotNull(userInfo, "UserInfo");
 		EvaluationUtils.ensureNotNull(evalId, "Evaluation ID");
 		Evaluation comp = getEvaluation(evalId);
-		return isEvalAdmin(userInfo, comp);
+		return EvaluationUtil.isEvalAdmin(userInfo, comp);
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/evaluation/manager/EvaluationManagerImpl.java:126]: can we move all authorization logic to the authorization manager.

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/Synapse.java [hunk 4]
```diff
@@ -1080,6 +1080,19 @@ public Annotations updateAnnotations(String entityId, Annotations updated) throw
 		}
 	}
 
+	public VariableContentPaginatedResults<AccessRequirement> getUnmetEvaluationAccessRequirements(String evaluationId) throws SynapseException {
```
Comment (john-hill) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/Synapse.java]: merge into a single method that takes a type like the wiki stuff.

File: services/repository-managers/src/main/java/org/sagebionetworks/evaluation/manager/ParticipantManagerImpl.java [hunk 5]
```diff
@@ -109,16 +119,16 @@ public void removeParticipant(UserInfo userInfo, String evalId, String idToRemov
 		String principalId = userInfo.getIndividualGroup().getId();
 		
 		// verify permissions
-		if (!EvaluationManager.isEvalAdmin(userInfo, evalId)) {
+		if (!evaluationManager.isEvalAdmin(userInfo, evalId)) {
 			// user is not an admin; only authorized to cancel their own participation
-			EvaluationUtils.ensureEvaluationIsOpen(EvaluationManager.getEvaluation(evalId));
+			EvaluationUtils.ensureEvaluationIsOpen(evaluationManager.getEvaluation(evalId));
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/evaluation/manager/ParticipantManagerImpl.java:124]: this call can be moved to the authorization manager(which can use the util)

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AccessApprovalManagerImpl.java [hunk 6]
```diff
@@ -88,11 +93,20 @@ public static void populateModifiedFields(UserInfo userInfo, AccessApproval a) {
...
 
 	@Override
-	public QueryResults<AccessApproval> getAccessApprovalsForEntity(
-			UserInfo userInfo, String entityId) throws DatastoreException,
+	public QueryResults<AccessApproval> getAccessApprovalsForSubject(
+			UserInfo userInfo, RestrictableObjectDescriptor subjectId) throws DatastoreException,
 			NotFoundException, ForbiddenException {
-		ACTUtils.verifyACTTeamMembershipOrIsAdmin(userInfo, userGroupDAO);
-		List<AccessRequirement> ars = accessRequirementDAO.getForNode(entityId);
+		if (RestrictableObjectType.ENTITY.equals(subjectId.getType())) {
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AccessApprovalManagerImpl.java:99]: move this logic to authorization manager.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AccessRequirementManagerImpl.java [hunk 7]
```diff
@@ -73,34 +73,53 @@ public static void populateModifiedFields(UserInfo userInfo, AccessRequirement a
 	@Override
 	public <T extends AccessRequirement> T createAccessRequirement(UserInfo userInfo, T accessRequirement) throws DatastoreException, InvalidModelException, UnauthorizedException, NotFoundException, ForbiddenException {
 		validateAccessRequirement(accessRequirement);
-		ACTUtils.verifyACTTeamMembershipOrCanCreateOrEdit(
-				userInfo, 
-				accessRequirement.getEntityIds(),
-				userGroupDAO, 
-				authorizationManager);
+		 Map<RestrictableObjectType, Collection<String>> sortedIds =
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AccessRequirementManagerImpl.java:76]: move logic to auto-manager.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/AccessRequirementManagerImplAutoWiredTest.java [hunk 8]
```diff
@@ -35,16 +44,29 @@
...
 	@Autowired
 	public UserProvider testUserProvider;
+	
+	@Autowired
+	public UserGroupDAO userGroupDAO;
+	
 	@Autowired
 	public AccessRequirementManager accessRequirementManager;
 	
+	@Autowired
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/AccessRequirementManagerImplAutoWiredTest.java:55]: does the manager already do id and creation stuff for you?

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/AccessRequirementManagerImplAutoWiredTest.java [hunk 9]
```diff
@@ -81,24 +120,60 @@ public void after() throws Exception {
...
+			evaluation=null;
+		}
+		if (adminEvaluation!=null) {
+			evaluationDAO.delete(adminEvaluation.getId());
+			adminEvaluation=null;
+		}
 	}
 	
-	private static TermsOfUseAccessRequirement newAccessRequirement(String entityId) {
+	private static TermsOfUseAccessRequirement newEntityAccessRequirement(String entityId) {
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/AccessRequirementManagerImplAutoWiredTest.java:135]: is redundant with mocked tests?

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/AccessRequirementManagerImplAutoWiredTest.java [hunk 10]
```diff
@@ -106,84 +181,145 @@ public void testCreateAccessRequirement() throws Exception {
...
 	
 	@Test(expected=ForbiddenException.class)
-	public void testCreateAccessRequirementForbidden() throws Exception {
-		ar = newAccessRequirement(entityId);
+	public void testEntityCreateAccessRequirementForbidden() throws Exception {
+		ar = newEntityAccessRequirement(entityId);
+		ar = accessRequirementManager.createAccessRequirement(testUserProvider.getTestUserInfo(), ar);
+	}
+	
+//	private UserInfo createDummyUserInfo() {
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/AccessRequirementManagerImplAutoWiredTest.java]: remove dead code

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/AccessRequirementManagerImplAutoWiredTest.java [hunk 11]
```diff
@@ -106,84 +181,145 @@ public void testCreateAccessRequirement() throws Exception {
...
-		QueryResults<AccessRequirement> ars = accessRequirementManager.getAccessRequirementsForEntity(adminUserInfo, entityId);
+		RestrictableObjectDescriptor rod = new RestrictableObjectDescriptor();
+		rod.setId(entityId);
+		rod.setType(RestrictableObjectType.ENTITY);
+		QueryResults<AccessRequirement> ars = accessRequirementManager.getAccessRequirementsForSubject(adminUserInfo, rod);
 		assertEquals(0L, ars.getTotalNumberOfResults());
 		assertEquals(0, ars.getResults().size());
 	}
 	
+	@Test(expected=ForbiddenException.class)
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/AccessRequirementManagerImplAutoWiredTest.java]: Eliminate all ForbidenExceptions in code base.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/AuthorizationManagerImplTest.java [hunk 12]
```diff
@@ -585,4 +592,35 @@ public void testCanAccessActivityFail() throws Exception {
 		assertFalse(canAccess);
 	}	
 		
+//	public static Evaluation createNewEvaluation(String name, String ownerId) {
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/AuthorizationManagerImplTest.java]: dead code.

File: services/repository/src/main/java/org/sagebionetworks/repo/web/service/AccessApprovalService.java [hunk 13]
```diff
@@ -21,11 +22,16 @@ public AccessApproval createAccessApproval(String userId,
 			NotFoundException, ForbiddenException, InvalidModelException,
 			IOException;
 
-	public PaginatedResults<AccessApproval> getAccessApprovals(String userId,
+	public PaginatedResults<AccessApproval> getEntityAccessApprovals(String userId,
```
Comment (john-hill) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/service/AccessApprovalService.java]: the service can have one method for both that takes the type as a param.

File: services/repository/src/main/java/org/sagebionetworks/repo/web/service/AccessRequirementService.java [hunk 14]
```diff
@@ -14,16 +15,26 @@
...
 			throws DatastoreException, UnauthorizedException,
 			NotFoundException, ForbiddenException;
 
-	public PaginatedResults<AccessRequirement> getAccessRequirements(
+	public PaginatedResults<AccessRequirement> getEntityAccessRequirements(
 			String userId, String entityId,	HttpServletRequest request)
 			throws DatastoreException, UnauthorizedException,
 			NotFoundException, ForbiddenException;
 
+	public PaginatedResults<AccessRequirement> getUnfulfilledEvaluationAccessRequirements(
```
Comment (john-hill) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/service/AccessRequirementService.java]: same could have one method that takes a type.

File: services/repository/src/test/java/org/sagebionetworks/repo/web/controller/ServletTestHelper.java [hunk 15]
```diff
@@ -2204,5 +2259,61 @@ public static void removeFavorite(
 		return createPaginatedResultsFromJSON(response.getContentAsString(),
 				EntityHeader.class);
 	}
+	
+	public static Evaluation createEvaluation(
```
Comment (john-hill) [inline services/repository/src/test/java/org/sagebionetworks/repo/web/controller/ServletTestHelper.java:2263]: duplicate, use EntityServletTestHelper

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOSubjectAccessRequirement.java [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+
+	@Override
+	public MigratableTableTranslation<DBOSubjectAccessRequirement, DBONodeAccessRequirement> getTranslator() {
+
+		return new MigratableTableTranslation<DBOSubjectAccessRequirement, DBONodeAccessRequirement>(){
+
+			@Override
+			public DBOSubjectAccessRequirement createDatabaseObjectFromBackup(
+					DBONodeAccessRequirement backup) {
+				DBOSubjectAccessRequirement dbo = new DBOSubjectAccessRequirement();
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOSubjectAccessRequirement.java:150]: replace DBONodeAccessRequirement as a backup object with something that has the same fields but also supports the new fields of DBOSubjectAccessRequirement.

**Pull request merged**